### ISSUE TITLE
SSO Download Error & FOREIGN KEY constraint failed

### ISSUE THREAD

**Issue by xdubx** [stefan-niedermann/nextcloud-deck#1254]
**Describe the bug**
Deck App don't download the boards. And stuck by the third one. After that, the first error appears a few times and the last error appears in the last window.

1. Go to 'Select account '
2. Click on 'Active Session from the nextcloud app'
3. Click "ok"

**Expected behavior**
Should download all boards.

**Screenshots**
None

**Versions**
 - Nextcloud: 22.2.3
 - Nextcloud Deck: 1.5.6
 - Nextcloud Android: 3.19
 - Nextcloud Android Deck: 1.20.0

**Smartphone (please complete the following information):**
OS Version: 4.4.153-perf+(V12.0.2.0.PEIMIXM)
OS API Level: 28
Device: whyred
Manufacturer: Xiaomi
Model (and Product): Redmi Note 5 (whyred)
- App-Store:
  - [ ] Google Play Store
  - [ ] Google Play Store (Beta channel)
  - [x] F-Droid
  - [ ] Huawei AppGallery

**Stacktrace**

```
App Version: 1.20.0
App Version Code: 1020000
Server App Version: 1.5.6
App Flavor: fdroid

Files App Version Code: 30190090

---

OS Version: 4.4.153-perf+(V12.0.2.0.PEIMIXM)
OS API Level: 28
Device: whyred
Manufacturer: Xiaomi
Model (and Product): Redmi Note 5 (whyred)

---

android.database.sqlite.SQLiteConstraintException: FOREIGN KEY constraint failed (code 787 SQLITE_CONSTRAINT_FOREIGNKEY)
	at android.database.sqlite.SQLiteConnection.nativeExecuteForLastInsertedRowId(Native Method)
	at android.database.sqlite.SQLiteConnection.executeForLastInsertedRowId(SQLiteConnection.java:796)
	at android.database.sqlite.SQLiteSession.executeForLastInsertedRowId(SQLiteSession.java:788)
	at android.database.sqlite.SQLiteStatement.executeInsert(SQLiteStatement.java:86)
	at androidx.sqlite.db.framework.FrameworkSQLiteStatement.executeInsert(FrameworkSQLiteStatement.java:51)
	at androidx.room.EntityInsertionAdapter.insertAndReturnId(EntityInsertionAdapter.java:114)
	at it.niedermann.nextcloud.deck.persistence.sync.adapters.db.dao.StackDao_Impl.insert(StackDao_Impl.java:177)
	at it.niedermann.nextcloud.deck.persistence.sync.adapters.db.dao.StackDao_Impl.insert(StackDao_Impl.java:33)
	at it.niedermann.nextcloud.deck.persistence.sync.adapters.db.DataBaseAdapter.createStack(DataBaseAdapter.java:645)
	at it.niedermann.nextcloud.deck.persistence.sync.helpers.providers.StackDataProvider.createInDB(StackDataProvider.java:44)
	at it.niedermann.nextcloud.deck.persistence.sync.helpers.providers.StackDataProvider.createInDB(StackDataProvider.java:20)
	at it.niedermann.nextcloud.deck.persistence.sync.helpers.SyncHelper$1.onResponse(SyncHelper.java:62)
	at it.niedermann.nextcloud.deck.persistence.sync.helpers.SyncHelper$1.onResponse(SyncHelper.java:47)
	at it.niedermann.nextcloud.deck.api.RequestHelper$ResponseConsumer.accept(RequestHelper.java:56)
	at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:63)
	at io.reactivex.internal.operators.observable.ObservableSubscribeOn$SubscribeOnObserver.onNext(ObservableSubscribeOn.java:58)
	at io.reactivex.internal.operators.observable.ObservableMap$MapObserver.onNext(ObservableMap.java:62)
	at io.reactivex.internal.operators.observable.ObservableFromPublisher$PublisherSubscriber.onNext(ObservableFromPublisher.java:56)
	at com.nextcloud.android.sso.api.NextcloudAPI.lambda$performRequestObservableV2$1$com-nextcloud-android-sso-api-NextcloudAPI(NextcloudAPI.java:130)
	at com.nextcloud.android.sso.api.NextcloudAPI$$ExternalSyntheticLambda2.subscribe(Unknown Source:6)
	at io.reactivex.internal.operators.observable.ObservableFromPublisher.subscribeActual(ObservableFromPublisher.java:31)
	at io.reactivex.Observable.subscribe(Observable.java:12284)
	at io.reactivex.internal.operators.observable.ObservableMap.subscribeActual(ObservableMap.java:32)
	at io.reactivex.Observable.subscribe(Observable.java:12284)
	at io.reactivex.internal.operators.observable.ObservableSubscribeOn$SubscribeTask.run(ObservableSubscribeOn.java:96)
	at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:38)
	at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:26)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:764)

```

```
App Version: 1.20.0
App Version Code: 1020000
Server App Version: 1.5.6
App Flavor: fdroid

Files App Version Code: 30190090

---

OS Version: 4.4.153-perf+(V12.0.2.0.PEIMIXM)
OS API Level: 28
Device: whyred
Manufacturer: Xiaomi
Model (and Product): Redmi Note 5 (whyred)

---

com.nextcloud.android.sso.exceptions.NextcloudHttpRequestFailedException: HTTP-Anfrage ist fehlgeschlagen mit HTTP-Statuscode: 403
	at com.nextcloud.android.sso.api.AidlNetworkRequest.performNetworkRequestV2(AidlNetworkRequest.java:188)
	at com.nextcloud.android.sso.api.NextcloudAPI.performNetworkRequestV2(NextcloudAPI.java:199)
	at com.nextcloud.android.sso.api.NextcloudAPI.lambda$performRequestObservableV2$1$com-nextcloud-android-sso-api-NextcloudAPI(NextcloudAPI.java:129)
	at com.nextcloud.android.sso.api.NextcloudAPI$$ExternalSyntheticLambda2.subscribe(Unknown Source:6)
	at io.reactivex.internal.operators.observable.ObservableFromPublisher.subscribeActual(ObservableFromPublisher.java:31)
	at io.reactivex.Observable.subscribe(Observable.java:12284)
	at io.reactivex.internal.operators.observable.ObservableMap.subscribeActual(ObservableMap.java:32)
	at io.reactivex.Observable.subscribe(Observable.java:12284)
	at io.reactivex.internal.operators.observable.ObservableSubscribeOn$SubscribeTask.run(ObservableSubscribeOn.java:96)
	at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:38)
	at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:26)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:764)
Caused by: java.lang.IllegalStateException: {"status":403,"message":"Permission denied"}

	at com.nextcloud.android.sso.InputStreamBinder.processRequestV2(InputStreamBinder.java:455)
	at com.nextcloud.android.sso.InputStreamBinder.performNextcloudRequestAndBodyStreamV2(InputStreamBinder.java:128)
	at com.nextcloud.android.sso.InputStreamBinder.performNextcloudRequestV2(InputStreamBinder.java:111)
	at com.nextcloud.android.sso.aidl.IInputStreamService$Stub.onTransact(IInputStreamService.java:158)
	at android.os.Binder.execTransact(Binder.java:735)

```

I cannot explain the 403 error. I use Nextcloud Talk and the Nextcloud app without problems with Keycloak as SSO. Maybe someone has a tip on what I can do to fix the error?

**Comment by stefan-niedermann** [comment]
@desperateCoder I think you already figured that the `SQLConstraint` exception was caused by the used `ThreadPoolExecutor`s, no? Did you already push your fix into `master` and do we want to take https://github.com/stefan-niedermann/nextcloud-deck/issues/1210 (with branch [`1210-fixed-thread-pools`](https://github.com/stefan-niedermann/nextcloud-deck/tree/1210-fixed-thread-pools) - still needs some work though, I think)?

For refetence: this issue has also been [reported at help.nextcloud com](https://help.nextcloud.com/t/unable-to-sync-deck-on-initial-setup-android-client/137325).

**Comment by xdubx** [comment]
I dated my nextcloud version to v24 and the deck app to and the error is gone. Meanwhile I changed my is version to android 11.
Someone of this factors solve the problem.

**Comment by stefan-niedermann** [comment]
This cause of this issue should be fixed in `1.20.1`. Please reopen in case the problem occurs again with this or a higher version.

Thank you for the report!
